src/lib/functions/netlify-function.js

```diff
@@ -0,0 +1,104 @@
+const BACKGROUND_SUFFIX = '-background'
+
+class NetlifyFunction {
+  constructor({
+    config,
+    errorExit,
+    functionsDirectory,
+    mainFile,
+    name,
+    projectRoot,
+    runtime,
+    timeoutBackground,
+    timeoutSynchronous,
+  }) {
+    this.config = config
+    this.errorExit = errorExit
+    this.functionsDirectory = functionsDirectory
+    this.mainFile = mainFile
+    this.name = name
+    this.projectRoot = projectRoot
+    this.runtime = runtime
+    this.timeoutBackground = timeoutBackground
+    this.timeoutSynchronous = timeoutSynchronous
+
+    // Determines whether this is a background function based on the function
+    // name.
+    this.isBackground = name.endsWith(BACKGROUND_SUFFIX)
+
+    // List of the function's source files. This starts out as an empty set
+    // and will get populated on every build.
+    this.srcFiles = new Set()
+  }
+
+  async build() {
+    const buildFunction = await this.runtime.getBuildFunction({
+      config: this.config,
+      errorExit: this.errorExit,
+      func: this,
+      functionsDirectory: this.functionsDirectory,
+      projectRoot: this.projectRoot,
+    })
+
+    this.buildQueue = typeof buildFunction === 'function' ? buildFunction(this) : undefined
+
+    try {
+      const { srcFiles, ...buildData } = (await this.buildQueue) || {}
+      const srcFilesSet = new Set(srcFiles)
+      const srcFilesDiff = this.getSrcFilesDiff(srcFilesSet)
+
+      this.buildData = buildData
+      this.srcFiles = srcFilesSet
+
+      return { srcFilesDiff }
+    } catch (error) {
+      return { error }
+    }
+  }
+
+  getSrcFilesDiff(srcFiles) {
+    const added = new Set()
+    const deleted = new Set()
+
+    srcFiles.forEach((path) => {
+      if (!this.srcFiles.has(path)) {
+        added.add(path)
+      }
+    })
+
+    this.srcFiles.forEach((path) => {
+      if (!srcFiles.has(path)) {
+        deleted.add(path)
+      }
+    })
+
+    return {
+      added,
+      deleted,
+    }
+  }
+
+  async invoke(event, context) {
+    await this.buildQueue
```

Comment:
If the function is currently being built, we wait.

This means that the Dev server is able to start much quicker than it does currently, because we no longer have to wait for all functions to be built (e.g. using `zip-it-and-ship-it`) before we start accepting requests. Instead, we build the functions in the background and if a request happens to come in while a function is still building, we wait for the build to finish before serving the request.